ipa-multipoint/src/main/java/org/hyperledger/besu/nativelib/ipamultipoint/LibIpaMultipoint.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Besu Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.nativelib.ipamultipoint;
+
+/**
+ * Java interface to ipa-multipoint.
+ *
+ * Allows to compute commitments for Verkle tree nodes.
+ */
+public class LibIpaMultipoint {
+
+  public static native byte[] commit(byte[] input1, byte[] input2, byte[] input3, byte[] input4);
```

Comment:
Are we always only committing 4 values?  Do each of the values have meaning? I think a bit more javadoc would be appropriate if it is going into a general release.

Reply:
1. yes, always 4 values.
2. They form a lagrange basis point - and that's all I know. @gballet can you help document?

I'm not aware of a draft spec for ipa multipoint, I just built this on top of the rust lib as you can see, and it's not documenting this either. I took the route of admitting that I had no idea what I was looking at (I'm really not a mathematician) and that I'd make it work, but sure, it'd be nice to know more about the ins and outs.

Reply:
This is a crypto library.  We should either be able to explain it or point to papers/libraries/written material that explain it.  How can we even be sure we are using the crypto primitive correctly?

Reply:
Pretty much, I took the test vector in rust and made it work for Java IIRC.

Reply:
Some clarifications here:

1. the curve is called bandersnatch and is described here https://eprint.iacr.org/2021/1152.pdf
2. this `commit` function is simply the evaluation of a polynomial at a specific point. This is what polynomial commitments mean: the evaluation of a polynomial at a given point, that you can later use to prove that you know the full polynomial.
3. In the context of verkle trees, there are two such polynomial evaluations that can occur:
  * When calculating a so-called "pedersen hash", in which a degree 3 polynomial is evaluated
  * When calculating the commitment to a tree node, in which a degree 255 polynomial is evaluated

This first function therefore corresponds to the first case: the evaluation of a polynomial of degree 3 (uniquely defined by 4 values). For your documentation effort, maybe the function should be renamed to something like `commitToDegree3Polynomial` and the 4 values be renamed `y0`, `y1`, `y2` and `y3` as they are the y values that this polynomial takes at x=0,1,2,3.

Alternatively, we could imagine modifying these values to take a polynomial of any degree, but I don't think that it's possible for JNI to pass `[][]byte` values, is it?

Reply:
We could change the Java signature, but for now let's address the problem of the missing documentation. I am going to take your comments and turn them into code changes, such as Java comments, to help complete the Java method signature and get us past this current hurdle.